drivers/codec/wm8960.py

```diff
@@ -0,0 +1,940 @@
+
+_WM8960_I2C_ADDR = const(0x1A)
+
+# WM8960 maximum volume values
+_WM8960_ADC_MAX_VOLUME = const(0xFF)
+_WM8960_DAC_MAX_VOLUME = const(0xFF)
+_WM8960_HEADPHONE_MAX_VOLUME = const(0x7F)
+_WM8960_LINEIN_MAX_VOLUME = const(0x3F)
+_WM8960_SPEAKER_MAX_VOLUME = const(0x7F)
+
+# Config symbol names
+# Modules
+module_ADC = const(0)  # ADC module in WM8960
+module_DAC = const(1)  # DAC module in WM8960
+module_VREF = const(2)  # VREF module
+module_headphone = const(3)  # Headphone
+module_mic_bias = const(4)  # Mic bias
+module_mic = const(5)  # Input Mic
+module_line_in = const(6)  # Analog in PGA
+module_line_out = const(7)  # Line out module
+module_speaker = const(8)  # Speaker module
+module_omix = const(9)  # Output mixer
+module_mono_out = const(10)  # Mono mix
+
+# Route
+route_bypass = const(0)  # LINEIN->Headphone.
+route_playback = const(1)  #  I2SIN->DAC->Headphone.
+route_playback_record = const(2)  # I2SIN->DAC->Headphone, LINEIN->ADC->I2SOUT.
+route_record = const(5)  # LINEIN->ADC->I2SOUT.
+
+# Input
+input_closed = const(0)  # Input device is closed
+input_mic1 = const(1)  # Input as single ended mic, only use L/RINPUT1
+input_mic2 = const(2)  # Input as diff. mic, use L/RINPUT1 and L/RINPUT2
+input_mic3 = const(3)  # Input as diff. mic, use L/RINPUT1 and L/RINPUT3
+input_line2 = const(4)  # Input as line input, only use L/RINPUT2
+input_line3 = const(5)  # Input as line input, only use L/RINPUT3
+
+# ADC sync input
+sync_adc = const(0)  # Use ADCLRC pin for ADC sync
+sync_dac = const(1)  # used DACLRC pin for ADC sync
+
+# Protocol type
+bus_I2S = const(2)  # I2S type
+bus_left_justified = const(1)  # Left justified mode
+bus_right_justified = const(0)  # Right justified mode
+bus_PCMA = const(3)  # PCM A mode
+bus_PCMB = const(3 | (1 << 4))  # PCM B mode
+
+# Channel swap
+swap_none = const(0)
+swap_input = const(1)
+swap_output = const(2)
+
+# Mute settings
+mute_fast = const(0)
+mute_slow = const(1)
+
+# ALC settings
+alc_off = const(0)
+alc_right = const(1)
+alc_left = const(2)
+alc_stereo = const(3)
+alc_mode = const(0)  # ALC mode
+alc_limiter = const(1)  # Limiter mode
+
+# Clock Source
+sysclk_mclk = const(0)  # sysclk source from external MCLK
+sysclk_PLL = const(1)  # sysclk source from internal PLL
+
+
+class WM8960:
+
+    # register cache of 56 register. Since registers cannot be read back, they are
+    # kept in the table for modification
+    # fmt: off
+    reg = [
+        0x0097, 0x0097, 0x0000, 0x0000, 0x0000, 0x0008, 0x0000,
+        0x000a, 0x01c0, 0x0000, 0x00ff, 0x00ff, 0x0000, 0x0000,
+        0x0000, 0x0000, 0x0000, 0x007b, 0x0100, 0x0032, 0x0000,
+        0x00c3, 0x00c3, 0x01c0, 0x0000, 0x0000, 0x0000, 0x0000,
+        0x0000, 0x0000, 0x0000, 0x0000, 0x0100, 0x0100, 0x0050,
+        0x0050, 0x0050, 0x0050, 0x0000, 0x0000, 0x0000, 0x0000,
+        0x0040, 0x0000, 0x0000, 0x0050, 0x0050, 0x0000, 0x0002,
+        0x0037, 0x004d, 0x0080, 0x0008, 0x0031, 0x0026, 0x00e9
+    ]
```

Comment:
If possible, better to make this a tuple, then it has a chance to be optimised by the compiler.

Reply:
I'm writing back into that table. So a tuple is not feasible.